nixos/lib/test-driver/test_driver/driver.py

```diff
@@ -101,6 +110,153 @@ def __exit__(self, *_: Any) -> None:
+                        )
+                    else:
+                        to_del.append(idx)
+            for idx in sorted(to_del, reverse=True):
+                py_name = pythonize_name(self.machines[idx].name)
+                self.logger.info(
+                    f"{self.machines[idx].name} removed from the test. deleting it from the environment"
+                )
+                del repl_globals[py_name]
+                del self.machines[idx]
+
+            # add and change new machines
+            for start_cmd in start_cmds:
+                existing = [
+                    m for m in self.machines if m.name == start_cmd.machine_name
+                ]
+                if len(existing) == 0:
+                    machine = self.create_machine(start_cmd._cmd)
+                    py_name = pythonize_name(machine.name)
+                    repl_globals[py_name] = machine
+                    self.machines.append(machine)
+
+                    self.logger.info(
+                        f"new machine created {start_cmd.machine_name}. start it with {py_name}.start()"
+                    )
+                elif len(existing) == 1:
+                    machine = existing[0]
+                    if machine.start_command._cmd == start_cmd._cmd:
+                        self.logger.info(f"{start_cmd.machine_name} unchanged.")
+                    elif not machine.booted:
+                        self.logger.info(f"{start_cmd.machine_name} changed.")
+                        machine.start_command = start_cmd
+                    else:
+                        with self.logger.nested(
+                            f"{start_cmd.machine_name} updated. switching to new configuration"
+                        ):
+                            store_path = Path(start_cmd._cmd).parent.parent
+                            store_path_exists, _ = machine.execute(f"ls {store_path}")
+                            if store_path_exists != 0:
+                                self.logger.error(
+                                    f"Skipping the update of {start_cmd.machine_name}, because the store output {start_cmd._cmd} does not exist on the VM. This will happen if `virtualisation.useNixStoreImage = true;`."
+                                )
+                            else:
+                                switch_cmd = (
+                                    store_path
+                                    / "system"
+                                    / "bin"
+                                    / "switch-to-configuration"
+                                )
+                                machine.succeed(f"{switch_cmd} test")
+                                machine.start_command = start_cmd
+                else:
+                    self.logger.error(
+                        f"Skipping the update of {start_cmd.machine_name}, because there are mltiple machines with that name. This shouldn't be possible, and is an error in the testing framework."
```

Comment:
```suggestion
                        f"Skipping the update of {start_cmd.machine_name}, because there are multiple machines with that name. This shouldn't be possible, and is an error in the testing framework."
```